### ISSUE TITLE
Password authentication

### ISSUE THREAD

**Issue by kelleg1** [iaincollins/next-auth#9]
I'd like to use Next.js Starter to use email & password-based authentication. It seems next-auth takes care of this now as it adds the /email/signin route.
Is it possible to add a way to supply a username and password validation function to next-auth (that I can connect to MongoDB inside of)?

Something else to consider is using bcrypt to generate password hashes. I don't need this now since I will not be doing user registration (only login), but I might need it later on for other projects.

**Comment by iaincollins** [comment]
This is now resolved in 1.8.0, which allows any authentication mechanism by defining a custom signIn() function which enables a POST based route to /auth/signin.

As this approach doesn't it prescribe the fields it can be used for username and password, two factor authentication or even token based SSO integration. An example of how to configure and use it provided in the examples folder.

**Comment by onomatopoetry** [comment]
Hello, I am trying to conduct a very simple test of the custom signIn() function. I've downloaded the example from github, run `npm install` in the 'example' directory, then `npm run dev`. Without modifying anything else in the example, I've uncommented the signIn() function, and altered it like so:
```
signIn: ({form, req}) => {
  return new Promise((resolve, reject) => {
    // Should validate credentials (e.g. hash password, compare 2FA token
    // etc) and return a valid user object from a database.
    /*
      return usersCollection.findOne({
      email: form.email
    }, (err, user) => {
      if (err) return reject(err)
      if (!user) return resolve(null)
      
      // Check credentials - e.g. compare bcrypt password hashes
      if (form.password === "test1234") {
        // If valid, return user object - e.g. { id, name, email }
        return resolve(user)
      } else {
        // If invalid, return null
        return resolve(null)
      }
    })
    */
    resolve({id:1234, name: 'Test', email: 'test@gmail.com'})
  })
}
```
Now when I visit http://localhost:3000/auth/credentials and input anything into the "Email Address" and "Password" fields, I appear to login successfully. I see in the network tab that /auth/signin is being visited, followed by the callback.js page. I'm then returned to the index page _but i am not signed in._ The session object does not include a user object, and the "Manage Account" and "Sign out" links are not appearing.

What am I doing wrong?
